src/dotnet/CarbonAware/IntervalHelper.cs

```diff
@@ -0,0 +1,66 @@
+namespace CarbonAware;
+
+/// <summary>
+/// Helper for intervals within Carbon Aware.
+/// </summary>
+public static class IntervalHelper
+{
+
+    /// <summary>
+    /// Ensures the data is valid by using the min sampling window. Uses the expanded data to add an extra
+    /// data point before the startDate to ensure all expected data is included.
+    /// </summary>
+    /// <param name="data">Expanded data that includes the minimum sampling window</param>
+    /// <param name="startDate">Original start date provided by user</param>
+    /// <param name="endDate">Original end date provided by user</param>
+    /// <returns>Filtered emissions data.</returns>
+    /// <remarks> If the user provided interval is already >= minimum sampling window, or no expanded data, we just return the passed data. </remarks>
+    public static IEnumerable<EmissionsData> MinSamplingFiltering(IEnumerable<EmissionsData> expandedData, DateTimeOffset startDate, DateTimeOffset endDate, double minSamplingWindow)
+    {
+        TimeSpan ts = endDate - startDate;
+        if ((ts.TotalMinutes >= minSamplingWindow) || !expandedData.Any())
+        {
+            return expandedData;
+        }
+
+        var arrData = expandedData.ToArray();
+        var startDateTime = startDate.DateTime;
+        var dataLength = arrData.Length;
+        var splitIndex = dataLength - 1;
+
+        // sort data since different sources might have populated the data differently.
+        Array.Sort(arrData, new CompareEmissionDataSort());
```

Comment:
Given that we sort through all of the data anyway, could we also use that pass through the data to also filter out the unneeded datapoints?

Reply:
afaik it won't be possible since sorting follows the framework based on a comparer class, which the only thing is doing is organizing data, not filtering, by 'removing' items. This has to happen outside of Array.Sort.